### PR TITLE
Feature 58 products calculated price

### PR DESCRIPTION
**Pull request by berki0**
We need to change the Resources prices to be able to create a Sale. (The reason I didn't change them via flyway migration is that the updateResourceTest was not saved in System Events and we were getting an error)

### CODE REVIEW

File: src/main/java/jewellery/inventory/calculator/ProductPriceCalculator.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/calculator/ProductPriceCalculator.java:1]: I am not a fan of the calculator package name. Why not keep it in the ProductMapper? It's the only place it's used.

File: src/main/java/jewellery/inventory/calculator/ProductPriceCalculator.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/calculator/ProductPriceCalculator.java]: I don't think we need the annotations. The class has only static methods and doesn't store any other information.

File: src/main/java/jewellery/inventory/dto/request/ProductPriceDiscountRequestDto.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/request/ProductPriceDiscountRequestDto.java]: Maybe rename to `ProductDiscountRequestDto` now that it doesn't hold the price anymore?

File: src/main/java/jewellery/inventory/dto/request/ProductRequestDto.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/request/ProductRequestDto.java]: `additionalPrice` can be negative also. We don't have other constraints apart from NotNull maybe.

File: src/main/java/jewellery/inventory/mapper/ProductMapper.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ProductMapper.java]: It's not okay to call repository methods in mappers because it breaches separation of concerns. Mappers should have all the information to do the mapping without querying the database.

File: src/main/java/jewellery/inventory/mapper/ProductMapper.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ProductMapper.java]:
```suggestion
      productResponseDto.setSalePrice(
          product.getPartOfSale().getProducts().stream()
              .filter(
                  productPriceDiscount ->
                      product.getId().equals(productPriceDiscount.getProduct().getId()))
              .findFirst()
              .get()
              .getSalePrice());
```

File: src/main/java/jewellery/inventory/model/Product.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/model/Product.java]: do we need the column annotation? I see we haven't added it for the other columns

File: src/main/java/jewellery/inventory/model/Sale.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/model/Sale.java]: If we have CascadeType.REMOVE here do we need ProductPriceDiscountRepository and Service to delete the ProductPriceDiscount in the product return functionality

File: src/test/java/jewellery/inventory/helper/ProductPriceDiscountTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/ProductPriceDiscountTestHelper.java:1]: This looks like a method needed only for Sale related logic. Should we put the method in SaleTestHelper?

File: src/test/java/jewellery/inventory/helper/ProductTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/ProductTestHelper.java]: Pls, no.

File: src/test/java/jewellery/inventory/helper/SaleTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/SaleTestHelper.java]: This seems odd and obsolete. We are setting the products after the below foreach. Also why are we overwriting the sale's products here? It should already have products. Otherwise it's not a meaningful sale.

File: src/test/java/jewellery/inventory/helper/SaleTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/SaleTestHelper.java]: Seems this is not used anywhere

File: src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java]: This shouldn't be necessary.

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java]: We have a util class for generating BigDecimals. We can utilize it here

File: src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java]: there should be space after the comma here. My sight is turning bad and I thought this is a point 😢 Please, format the code.

File: src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java:102]: What made this assertion to change? What are we testing here? We should be more descriptive in the test name with the conditions we are testing.

File: src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java]: We have this repeating in every test here? This needs to be refactored to be part of the setup. From what I see in the previous state, the products were passed in the sale creation. I don't understand why we moved away from it?

**Review by VladoKat — CHANGES_REQUESTED** [review]

File: src/main/java/jewellery/inventory/model/Product.java
Comment (berki0) [reply src/main/java/jewellery/inventory/model/Product.java]: we need an annotation, if we don't have one, this field is saved in the Database as - (additionalprice)

File: src/main/java/jewellery/inventory/model/Sale.java
Comment (berki0) [reply src/main/java/jewellery/inventory/model/Sale.java]: we need functionality to delete ProductPriceDiscount because we can have in one Sale more than one ProductPriceDiscount and we can return only one from all from a given Sale, but if we delete a Sale this guarantees us that all ProductPriceDiscounts to it will be deleted

File: src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java
Comment (berki0) [reply src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java]: we need it because then we try to delete tables in the ProductRepository but we get an error(delete on table "product" violates foreign key constraint "product_price_discount_product_id_fkey" on table "product_price_discount")

File: src/test/java/jewellery/inventory/helper/ProductTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/ProductTestHelper.java]: Please, don't leave any commented code behind

File: src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleCrudIntegrationTest.java]: Another unformatted line. Please, make sure we format the code after we've made changes to it.

**Review by VladoKat — CHANGES_REQUESTED** [review]

File: src/main/java/jewellery/inventory/dto/request/ProductRequestDto.java
Comment (Alekseeybg) [inline src/main/java/jewellery/inventory/dto/request/ProductRequestDto.java]: Remove empty line

File: src/main/java/jewellery/inventory/mapper/ProductMapper.java
Comment (Alekseeybg) [inline src/main/java/jewellery/inventory/mapper/ProductMapper.java]: Remove empty line

File: src/main/java/jewellery/inventory/mapper/SaleMapper.java
Comment (Alekseeybg) [inline src/main/java/jewellery/inventory/mapper/SaleMapper.java]: remove empty line

File: src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java
Comment (Alekseeybg) [inline src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java]: remove empty line

File: src/main/java/jewellery/inventory/service/ProductPriceDiscountService.java
Comment (Alekseeybg) [inline src/main/java/jewellery/inventory/service/ProductPriceDiscountService.java]: remove empty line

File: src/main/java/jewellery/inventory/model/Product.java
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/model/Product.java]: We tested it and we found the problem being in the migrations. After the fix it will work as intended without the annotation.

File: src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java]: We are setting the products of the sale to be List.of(productPriceDiscount). There is no need to pass it in SaleTestHelper.getSaleResponseDto(sale, productPriceDiscount).

File: src/test/java/jewellery/inventory/integration/ResourceCrudIntegrationTest.java
Comment (Alekseeybg) [inline src/test/java/jewellery/inventory/integration/ResourceCrudIntegrationTest.java]: Reformat the file

File: src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/unit/mapper/SaleMapperTest.java:102]: The exception change comes from the change from double to BigDecimal. And due to changes in logic in calculations now we need to change the exception in the test.

File: src/test/java/jewellery/inventory/unit/service/ProductPriceDiscountServiceTests.java
Comment (Alekseeybg) [inline src/test/java/jewellery/inventory/unit/service/ProductPriceDiscountServiceTests.java:1]: Reformat the whole file

File: src/main/java/jewellery/inventory/mapper/ProductMapper.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ProductMapper.java]: should we extract this into getPriceFromSale() method?

File: src/main/java/jewellery/inventory/service/ProductPriceDiscountService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ProductPriceDiscountService.java]: Do we need to pass both Sale and saleRequestDto? We should be able to get all the information from Sale.

File: src/test/java/jewellery/inventory/helper/SaleTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/SaleTestHelper.java:1]: This file is not formatted.

File: src/test/java/jewellery/inventory/helper/SaleTestHelper.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/helper/SaleTestHelper.java]: We shouldn't pass the ProductPriceDiscount in the method as a separate argument. Sale has a list of ProductPriceDiscount inside it. The totalPrice should be the sum of all of the Products of the Sale. Similarly, discounttedPrice should be calculated the totalDisocunt and totalDiscountedPrice.

File: src/test/java/jewellery/inventory/unit/service/ProductPriceDiscountServiceTests.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ProductPriceDiscountServiceTests.java]: why are we using lenient here? I am not familiar with it.

File: src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java
Comment (VladoKat) [reply src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java]: Put the ProductPriceDiscount in the sale creation method -> `SaleTestHelper.createSaleWithTodayDate(seller, buyer);` add the product as an argument there and prepare it there. Then you would be able to create the saleResponseDto from the sale directly.

File: src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java]: Why did we change this test? It no longer tests what happens with a sale when there are multiple (2) products inside it.

File: src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/SaleServiceTest.java:196]: This looks suspicious because we are expecting the sale to not be deleted after returning 1 out of 2 products.

**Review by VladoKat — CHANGES_REQUESTED** [review]

File: src/main/java/jewellery/inventory/mapper/ProductMapper.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ProductMapper.java]: It seems to me we shouldn't need to do the second check `!product.getPartOfSale().getProducts().isEmpty()`. If the product is part of a sale, then surely the products of the sale shouldn't be an empty array.

File: src/main/java/jewellery/inventory/model/Sale.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/model/Sale.java:1]: Seems file is not formatted.

File: src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java:1]: Do we still need this repository?

File: src/main/java/jewellery/inventory/service/ProductService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ProductService.java]: Let's make `ProductMapper::calculateTotalPrice(Product product)` public and use it instead of mapping the whole product just to get the price.

File: src/main/java/jewellery/inventory/service/SaleService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/SaleService.java]: This method would look a bit better if we use sale.getProducts().forEach(....). Or alternatively, extract sale.getProducts().get(i) as currentProductPriceDiscount to not have so many repetitions of sale.getProducts().get(i)

File: src/main/resources/db/migration/dev/V1_7__create_product_price_discount_Table.sql
Comment (VladoKat) [inline src/main/resources/db/migration/dev/V1_7__create_product_price_discount_Table.sql:1]: We should have this migration in migration/prod as well

**Review by VladoKat — CHANGES_REQUESTED** [review]
Another set of minor change requests.

File: src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java
Comment (berki0) [reply src/main/java/jewellery/inventory/repository/ProductPriceDiscountRepository.java:1]: We use them in **AuthenticateIntegrationTestBase** for deleting ProductPriceDiscount tables because we have foreign key constraint to Product

File: src/main/java/jewellery/inventory/service/ProductService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ProductService.java:82]: This method should either return the calculated price or the price from the sale depending on whether the product is sold or not.

File: src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/AuthenticatedIntegrationTestBase.java]: Is this an obsolete import? I guess we are not using it anywhere.

**Review by VladoKat — CHANGES_REQUESTED** [review]
Last 2 minor changes. Will approve after that. Right now we have a bit of naming inconsistency though. Product has partOfSale that is of type ProductPriceDiscount. And Sale has products that is a list of ProductPriceDiscount... I can't think of a better naming for the fields and/or the ProductPriceDiscount. Could be ProductPriceDiscount becomes ProductInSale and then the names partOfSale in Product and products in Sale make more sense maybe. Think about it but I don't think it'd be a big improvement so I'll leave it up to you.

File: src/main/java/jewellery/inventory/service/ProductService.java
Comment (berki0) [reply src/main/java/jewellery/inventory/service/ProductService.java:82]: We use this method only when we create a Sale and for this we shouldn't need to check whether this Product has been sold because we have validations in CreateSale method

**Review by VladoKat — APPROVED** [review]

**Pull request merged**
